src/PROPOSAL/detail/PROPOSAL/Propagator.cxx

```diff
@@ -139,93 +142,112 @@ Interaction::Loss Propagator::DoStochasticInteraction(ParticleState& p_cond,
-                E_f = energy_next_interaction;
-            } else {
-                advance_grammage = density->Calculate(
-                        state.position, state.direction, advance_distance);
-                E_f = utility.EnergyDistance(state.energy, advance_grammage);
-            }
+    // Calculate maximal allowed length of step (limit due to final_distance)
+    const double max_distance = final_distance - state.propagated_distance;
 
-            std::tie(mean_direction, new_direction) = utility.DirectionsScatter(
-                advance_grammage, state.energy, E_f, state.direction, rnd);
+    // Calculate grammage until next stochastic interaction
+    double grammage_next_interaction = utility.LengthContinuous(
+            state.energy, energy_next_interaction);
 
+    int advancement_type;
+    Cartesian3D mean_direction, new_direction; // proposed scattering
+
+    // This lambda expression ensures we always use the same 4 random numbers
+    std::array<double, 4> random_numbers = {-1, -1, -1, -1};
+    int i = 0;
+    auto rnd = [&rnd_generator, &i, &random_numbers]() {
+        if (random_numbers[i%4] == -1)
+            random_numbers[i%4] = rnd_generator();
+        return random_numbers[i++%4];
+    };
+
+    do {
+        // Calculate grammage, energy and distance for step
+        if (energy != -1 && grammage == -1 && distance == -1) {
+            // Calculate grammage and distance from given energy
+            grammage = utility.LengthContinuous(state.energy, energy);
             try {
-                AdvanceDistance[ReachedInteraction]
-                    = density->Correct(state.position, mean_direction,
-                        grammage_next_interaction, max_distance_left);
+                distance = density->Correct(state.position, state.direction, grammage, max_distance);
             } catch (const DensityException&) {
-                AdvanceDistance[ReachedInteraction] = INF;
+                distance = INF;
+            }
+        } else if (energy == -1 && grammage == -1 && distance != -1) {
+            // Calculate energy and grammage from given distance
+            auto grammage_step = density->Calculate(state.position, state.direction, distance);
+            if (grammage_step < grammage_next_interaction) {
+                grammage = grammage_step;
+                energy = utility.EnergyDistance(state.energy, grammage);
+            } else {
+                // we are unable to reach `distance` before we reach the next interaction
+                // this means we are stuck in a loop, and need to discard the current set of random numbers
+                for (auto& r: random_numbers) {
+                    r = rnd_generator();
+                }
+                Logging::Get("proposal.propagator")->debug("Discard current set of random numbers for multiple"
+                                                           " scattering since we are stuck in a loop.");
```

Comment:
This can be more descriptive. Explain what the loop is.

Reply:
I've added a comment of the loop to describe shortly what it does.
And I also added a bit more context to describe why the random numbers are discarded in this case.